sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,253 @@
+import reactor.core.publisher.Sinks;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class ProactiveOpenConnectionsProcessor implements Closeable {
+
+    private static final Logger logger = LoggerFactory.getLogger(ProactiveOpenConnectionsProcessor.class);
+    private Sinks.Many<OpenConnectionTask> openConnectionsTaskSink;
+    private Sinks.Many<OpenConnectionTask> openConnectionsTaskSinkBackUp;
+    private final ConcurrentHashMap<String, List<OpenConnectionTask>> endpointsUnderMonitorMap;
+    private final AtomicReference<AsyncDocumentClient.OpenConnectionAggressivenessHint> aggressivenessHint;
+    private static final Map<AsyncDocumentClient.OpenConnectionAggressivenessHint, ConcurrencyConfiguration> concurrencySettings = new HashMap<>();
+    private final IOpenConnectionsHandler openConnectionsHandler;
+    private final RntbdEndpoint.Provider endpointProvider;
+    private Disposable openConnectionBackgroundTask;
+
+    static {
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.DEFENSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsDefensiveConcurrency(), Configs.getOpenConnectionsDefensiveConcurrency()));
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsAggressiveConcurrency(), Configs.getOpenConnectionsAggressiveConcurrency()));
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider, Duration aggressiveConnectionEstablishmentDuration) {
+        this.openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer();
+        this.openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer();
+        this.aggressivenessHint = new AtomicReference<>(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE);
+        this.endpointsUnderMonitorMap = new ConcurrentHashMap<>();
+        this.openConnectionsHandler = new RntbdOpenConnectionsHandler(endpointProvider);
+        this.endpointProvider = endpointProvider;
+
+        // start as a background task
+        openConnectionBackgroundTask = this.getOpenConnectionsPublisher();
+        Flux<Integer> backgroundOpenConnectionsSinkReInstantiationTask = Flux
+                .just(1)
+                .delayElements(aggressiveConnectionEstablishmentDuration);
+
+        if (aggressiveConnectionEstablishmentDuration != null && aggressiveConnectionEstablishmentDuration.compareTo(Duration.ZERO) > 0) {
+            backgroundOpenConnectionsSinkReInstantiationTask
+                    .doOnComplete(() -> reInstantiateOpenConnectionsPublisherAndSubscribe())
```

Comment:
all of the reInstantiate logic is happening still inside the constructor, while the object of this class is still being constructed. And it is using all the instance fields and mutating them in this logic. We should re-think on how to do this outside of the constructor, may be factory design pattern. 

A good read here - https://stackoverflow.com/questions/18348797/why-is-it-considered-bad-practice-in-java-to-call-a-method-from-within-a-constru

Reply:
Moved to the logic to an `init()` method instead.